database/src/ops/output.rs

```diff
@@ -1,34 +1,259 @@
 //! Outputs.
 
 //---------------------------------------------------------------------------------------------------- Import
+use monero_serai::transaction::{Timelock, Transaction};
 
-//---------------------------------------------------------------------------------------------------- Free Functions
-/// TODO
-pub fn add_output() {
-    todo!()
+use cuprate_types::{OutputOnChain, VerifiedBlockInformation};
+
+use crate::{
+    database::{DatabaseIter, DatabaseRo, DatabaseRw},
+    env::EnvInner,
+    error::RuntimeError,
+    ops::macros::{doc_add_block_inner_invariant, doc_error},
+    tables::{
+        BlockBlobs, BlockHeights, BlockInfos, KeyImages, NumOutputs, Outputs, PrunableHashes,
+        PrunableTxBlobs, PrunedTxBlobs, RctOutputs, Tables, TablesMut, TxHeights, TxIds,
+        TxUnlockTime,
+    },
+    transaction::{TxRo, TxRw},
+    types::{
+        Amount, AmountIndex, BlockHash, BlockHeight, BlockInfo, KeyImage, Output, PreRctOutputId,
+        RctOutput, TxHash,
+    },
+};
+
+//---------------------------------------------------------------------------------------------------- Pre-RCT Outputs
+/// Add a Pre-RCT [`Output`] to the database.
+///
+/// Upon [`Ok`], this function returns the [`PreRctOutputId`] that
+/// can be used to lookup the `Output` in [`get_output()`].
+///
+#[doc = doc_add_block_inner_invariant!()]
+#[doc = doc_error!()]
+#[inline]
+pub fn add_output(
+    amount: Amount,
+    output: &Output,
+    tables: &mut impl TablesMut,
+) -> Result<PreRctOutputId, RuntimeError> {
+    let amount_index = get_num_outputs(tables.outputs_mut())?;
+    tables.num_outputs_mut().put(&amount, &amount_index)?;
+
+    let pre_rct_output_id = PreRctOutputId {
+        amount,
+        amount_index,
+    };
+
+    tables.outputs_mut().put(&pre_rct_output_id, output)?;
+    Ok(pre_rct_output_id)
+}
+
+/// Remove a Pre-RCT [`Output`] from the database.
+#[doc = doc_add_block_inner_invariant!()]
+#[doc = doc_error!()]
+#[inline]
+pub fn remove_output(
+    pre_rct_output_id: &PreRctOutputId,
+    tables: &mut impl TablesMut,
+) -> Result<(), RuntimeError> {
+    // Decrement the amount index by 1, or delete the entry out-right.
+    tables
+        .num_outputs_mut()
+        .update(&pre_rct_output_id.amount, |amount| {
+            if amount == 0 {
```

Comment:
Oops, this is the `amount_index` not `amount`, we should be removing if `0` since it starts at `0` right?

```rust
if amount_index == 0 {
    None
} else {
    Some(amount_index - 1)
}
```